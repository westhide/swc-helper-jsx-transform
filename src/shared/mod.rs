pub use transform::Transform;

mod transform;
