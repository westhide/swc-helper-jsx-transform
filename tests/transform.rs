#![feature(box_syntax)]
#![feature(box_patterns)]

use insta::assert_debug_snapshot;
use swc_core::ecma::{
    ast::{JSXElement, JSXFragment},
    parser::{Syntax, TsConfig},
    transforms::testing::Tester,
    visit::{as_folder, VisitMut},
};
use swc_helper_jsx_transform::shared::Transform;

const TSX_SYNTAX: Syntax = Syntax::Typescript(TsConfig {
    tsx: true,
    decorators: false,
    dts: false,
    no_early_errors: false,
});

macro_rules! snapshot {
    ($name:ident, $value:expr) => {
        assert_debug_snapshot!(stringify!($name), $value.transform())
    };
}

macro_rules! test {
    ($name:ident, $src:literal) => {
        #[test]
        fn $name() {
            struct TransformVisitor;

            impl VisitMut for TransformVisitor {
                fn visit_mut_jsx_element(&mut self, element: &mut JSXElement) {
                    snapshot!($name, element)
                }

                fn visit_mut_jsx_fragment(&mut self, fragment: &mut JSXFragment) {
                    snapshot!($name, fragment)
                }
            }

            Tester::run(|tester| {
                tester.apply_transform(as_folder(TransformVisitor), "test.tsx", TSX_SYNTAX, $src)
            });
        }
    };

    ($($name:ident : $src:literal),+ $(,)?) => {
        $(test!($name,$src);)+
    };

    ($($mod:ident: {
        $($name:ident : $src:literal),+ $(,)?
    }),+ $(,)?) => {
        $(
        #[allow(non_snake_case)]
        mod $mod {
            use super::*;
            $(test!($name,$src);)+
        }
        )+
    };
}

test!(

    Tag:{
         native_div: r#"<div></div>"#,
         extra_component: r#"<A></A>"#,
         extra_custom: r#"<tag-custom></tag-custom>"#,
         extra_member: r#"<A.b></A.b>"#,
    },
    AttrKey:{
        attr: r#"<div class="cls"></div>"#,
        event_prefix: r#"<div onClick="()=>{}"></div>"#,
        event_namespace: r#"<div on:click="()=>{}"></div>"#,
        namespace_attr: r#"<div ns:name="value"></div>"#,
        spread: r#"<div {...spread}></div>"#,
    },
    AttrValue:{
        lit: r#"<div class="1"></div>"#,
        const_undefined: r#"<div class={undefined}></div>"#,
        const_array: r#"<div class={[1,2]}></div>"#,
        const_obj: r#"<div class={{a:1}}></div>"#,
        expr: r#"<div class={a}></div>"#,
        element: r#"<div element=<div></div>></div>"#,
        element_container: r#"<div element={<div></div>}></div>"#,
        fragment: r#"<div fragment=<></>></div>"#,
        fragment_container: r#"<div fragment={<></>}></div>"#,
        empty: r#"<div class></div>"#,
        boolean_attribute: r#"<div checked></div>"#,
    },

    Element:{
        is_static: r#"
          <div class="cls1" style="background-color: red;">
            <div id="firstChild"></div>
          </div>
        "#,
    },
    Fragment:{
        children: r#"
          <>
            <div></div>
            <div></div>
          </>
        "#,
    },
    Text:{
        clean_text:r#"
            <div>
             text1
             <br/>
       text2
             </div>
        "#,
    },
);
