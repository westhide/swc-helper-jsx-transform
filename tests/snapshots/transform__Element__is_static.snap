---
source: tests/transform.rs
expression: element.transform()
---
Element {
    tag: Native(
        "div",
    ),
    attrs: [
        Attr {
            key: Attr(
                "class",
            ),
            value: Lit(
                Str(
                    Str {
                        span: Span {
                            lo: BytePos(
                                23,
                            ),
                            hi: BytePos(
                                29,
                            ),
                            ctxt: #0,
                        },
                        value: Atom('cls1' type=inline),
                        raw: Some(
                            "\"cls1\"",
                        ),
                    },
                ),
            ),
        },
        Attr {
            key: Attr(
                "style",
            ),
            value: Lit(
                Str(
                    Str {
                        span: Span {
                            lo: BytePos(
                                36,
                            ),
                            hi: BytePos(
                                60,
                            ),
                            ctxt: #0,
                        },
                        value: Atom('background-color: red;' type=dynamic),
                        raw: Some(
                            "\"background-color: red;\"",
                        ),
                    },
                ),
            ),
        },
    ],
    children: [
        Element(
            Element {
                tag: Native(
                    "div",
                ),
                attrs: [
                    Attr {
                        key: Attr(
                            "id",
                        ),
                        value: Lit(
                            Str(
                                Str {
                                    span: Span {
                                        lo: BytePos(
                                            82,
                                        ),
                                        hi: BytePos(
                                            94,
                                        ),
                                        ctxt: #0,
                                    },
                                    value: Atom('firstChild' type=dynamic),
                                    raw: Some(
                                        "\"firstChild\"",
                                    ),
                                },
                            ),
                        ),
                    },
                ],
                children: [],
                raw: JSXElement {
                    span: Span {
                        lo: BytePos(
                            74,
                        ),
                        hi: BytePos(
                            101,
                        ),
                        ctxt: #0,
                    },
                    opening: JSXOpeningElement {
                        name: Ident(
                            Ident {
                                span: Span {
                                    lo: BytePos(
                                        75,
                                    ),
                                    hi: BytePos(
                                        78,
                                    ),
                                    ctxt: #0,
                                },
                                sym: Atom('div' type=static),
                                optional: false,
                            },
                        ),
                        span: Span {
                            lo: BytePos(
                                74,
                            ),
                            hi: BytePos(
                                95,
                            ),
                            ctxt: #0,
                        },
                        attrs: [
                            JSXAttr(
                                JSXAttr {
                                    span: Span {
                                        lo: BytePos(
                                            79,
                                        ),
                                        hi: BytePos(
                                            94,
                                        ),
                                        ctxt: #0,
                                    },
                                    name: Ident(
                                        Ident {
                                            span: Span {
                                                lo: BytePos(
                                                    79,
                                                ),
                                                hi: BytePos(
                                                    81,
                                                ),
                                                ctxt: #0,
                                            },
                                            sym: Atom('id' type=static),
                                            optional: false,
                                        },
                                    ),
                                    value: Some(
                                        Lit(
                                            Str(
                                                Str {
                                                    span: Span {
                                                        lo: BytePos(
                                                            82,
                                                        ),
                                                        hi: BytePos(
                                                            94,
                                                        ),
                                                        ctxt: #0,
                                                    },
                                                    value: Atom('firstChild' type=dynamic),
                                                    raw: Some(
                                                        "\"firstChild\"",
                                                    ),
                                                },
                                            ),
                                        ),
                                    ),
                                },
                            ),
                        ],
                        self_closing: false,
                        type_args: None,
                    },
                    children: [],
                    closing: Some(
                        JSXClosingElement {
                            span: Span {
                                lo: BytePos(
                                    95,
                                ),
                                hi: BytePos(
                                    101,
                                ),
                                ctxt: #0,
                            },
                            name: Ident(
                                Ident {
                                    span: Span {
                                        lo: BytePos(
                                            97,
                                        ),
                                        hi: BytePos(
                                            100,
                                        ),
                                        ctxt: #0,
                                    },
                                    sym: Atom('div' type=static),
                                    optional: false,
                                },
                            ),
                        },
                    ),
                },
                is_static: true,
            },
        ),
    ],
    raw: JSXElement {
        span: Span {
            lo: BytePos(
                12,
            ),
            hi: BytePos(
                118,
            ),
            ctxt: #0,
        },
        opening: JSXOpeningElement {
            name: Ident(
                Ident {
                    span: Span {
                        lo: BytePos(
                            13,
                        ),
                        hi: BytePos(
                            16,
                        ),
                        ctxt: #0,
                    },
                    sym: Atom('div' type=static),
                    optional: false,
                },
            ),
            span: Span {
                lo: BytePos(
                    12,
                ),
                hi: BytePos(
                    61,
                ),
                ctxt: #0,
            },
            attrs: [
                JSXAttr(
                    JSXAttr {
                        span: Span {
                            lo: BytePos(
                                17,
                            ),
                            hi: BytePos(
                                29,
                            ),
                            ctxt: #0,
                        },
                        name: Ident(
                            Ident {
                                span: Span {
                                    lo: BytePos(
                                        17,
                                    ),
                                    hi: BytePos(
                                        22,
                                    ),
                                    ctxt: #0,
                                },
                                sym: Atom('class' type=static),
                                optional: false,
                            },
                        ),
                        value: Some(
                            Lit(
                                Str(
                                    Str {
                                        span: Span {
                                            lo: BytePos(
                                                23,
                                            ),
                                            hi: BytePos(
                                                29,
                                            ),
                                            ctxt: #0,
                                        },
                                        value: Atom('cls1' type=inline),
                                        raw: Some(
                                            "\"cls1\"",
                                        ),
                                    },
                                ),
                            ),
                        ),
                    },
                ),
                JSXAttr(
                    JSXAttr {
                        span: Span {
                            lo: BytePos(
                                30,
                            ),
                            hi: BytePos(
                                60,
                            ),
                            ctxt: #0,
                        },
                        name: Ident(
                            Ident {
                                span: Span {
                                    lo: BytePos(
                                        30,
                                    ),
                                    hi: BytePos(
                                        35,
                                    ),
                                    ctxt: #0,
                                },
                                sym: Atom('style' type=static),
                                optional: false,
                            },
                        ),
                        value: Some(
                            Lit(
                                Str(
                                    Str {
                                        span: Span {
                                            lo: BytePos(
                                                36,
                                            ),
                                            hi: BytePos(
                                                60,
                                            ),
                                            ctxt: #0,
                                        },
                                        value: Atom('background-color: red;' type=dynamic),
                                        raw: Some(
                                            "\"background-color: red;\"",
                                        ),
                                    },
                                ),
                            ),
                        ),
                    },
                ),
            ],
            self_closing: false,
            type_args: None,
        },
        children: [
            JSXText(
                JSXText {
                    span: Span {
                        lo: BytePos(
                            61,
                        ),
                        hi: BytePos(
                            74,
                        ),
                        ctxt: #0,
                    },
                    value: "\n\n            ",
                    raw: "\n\n            ",
                },
            ),
            JSXElement(
                JSXElement {
                    span: Span {
                        lo: BytePos(
                            74,
                        ),
                        hi: BytePos(
                            101,
                        ),
                        ctxt: #0,
                    },
                    opening: JSXOpeningElement {
                        name: Ident(
                            Ident {
                                span: Span {
                                    lo: BytePos(
                                        75,
                                    ),
                                    hi: BytePos(
                                        78,
                                    ),
                                    ctxt: #0,
                                },
                                sym: Atom('div' type=static),
                                optional: false,
                            },
                        ),
                        span: Span {
                            lo: BytePos(
                                74,
                            ),
                            hi: BytePos(
                                95,
                            ),
                            ctxt: #0,
                        },
                        attrs: [
                            JSXAttr(
                                JSXAttr {
                                    span: Span {
                                        lo: BytePos(
                                            79,
                                        ),
                                        hi: BytePos(
                                            94,
                                        ),
                                        ctxt: #0,
                                    },
                                    name: Ident(
                                        Ident {
                                            span: Span {
                                                lo: BytePos(
                                                    79,
                                                ),
                                                hi: BytePos(
                                                    81,
                                                ),
                                                ctxt: #0,
                                            },
                                            sym: Atom('id' type=static),
                                            optional: false,
                                        },
                                    ),
                                    value: Some(
                                        Lit(
                                            Str(
                                                Str {
                                                    span: Span {
                                                        lo: BytePos(
                                                            82,
                                                        ),
                                                        hi: BytePos(
                                                            94,
                                                        ),
                                                        ctxt: #0,
                                                    },
                                                    value: Atom('firstChild' type=dynamic),
                                                    raw: Some(
                                                        "\"firstChild\"",
                                                    ),
                                                },
                                            ),
                                        ),
                                    ),
                                },
                            ),
                        ],
                        self_closing: false,
                        type_args: None,
                    },
                    children: [],
                    closing: Some(
                        JSXClosingElement {
                            span: Span {
                                lo: BytePos(
                                    95,
                                ),
                                hi: BytePos(
                                    101,
                                ),
                                ctxt: #0,
                            },
                            name: Ident(
                                Ident {
                                    span: Span {
                                        lo: BytePos(
                                            97,
                                        ),
                                        hi: BytePos(
                                            100,
                                        ),
                                        ctxt: #0,
                                    },
                                    sym: Atom('div' type=static),
                                    optional: false,
                                },
                            ),
                        },
                    ),
                },
            ),
            JSXText(
                JSXText {
                    span: Span {
                        lo: BytePos(
                            101,
                        ),
                        hi: BytePos(
                            112,
                        ),
                        ctxt: #0,
                    },
                    value: "\n\n          ",
                    raw: "\n\n          ",
                },
            ),
        ],
        closing: Some(
            JSXClosingElement {
                span: Span {
                    lo: BytePos(
                        112,
                    ),
                    hi: BytePos(
                        118,
                    ),
                    ctxt: #0,
                },
                name: Ident(
                    Ident {
                        span: Span {
                            lo: BytePos(
                                114,
                            ),
                            hi: BytePos(
                                117,
                            ),
                            ctxt: #0,
                        },
                        sym: Atom('div' type=static),
                        optional: false,
                    },
                ),
            },
        ),
    },
    is_static: true,
}
